remote_write: add metric for unexpected metadata in populateV2TimeSeries

leegin opened the pull request:
This PR addresses the TODO comment in populateV2TimeSeries by implementing logging and metrics tracking when unexpected metadata entries are encountered in samples-only batches.

### Changes

- Added logging when metadata is found in populateV2TimeSeries with actionable error messages pointing to [TODO](https://github.com/prometheus/prometheus/blob/1c002c5669b9b960f3fcbf9dec909c29fbc3fd85/storage/remote/queue_manager.go#L1963)

- Added metric tracking via unexpectedMetadataTotal to provide observability.

- Added comprehensive test TestPopulateV2TimeSeries_UnexpectedMetadata to verify the new behavior.

### Approach: Logging vs Error Handling

I chose the logging approach over returning errors because:

- Non-blocking: Allows valid samples to be processed even when metadata is mixed in.

- Graceful degradation: Remote write continues functioning while providing visibility into the issue.

- Production-friendly: Avoids breaking entire batches.

### Testing

Added unit test that verifies metadata detection and counting. Test confirms that samples are processed normally while metadata is logged and tracked. Validates that the unexpectedMetadataTotal metric is incremented correctly

### Related Issues

Addresses TODO comment in populateV2TimeSeries

Related to [14405](https://github.com/prometheus/prometheus/issues/14405)

### Open Questions

- Is the unexpectedMetadataTotal metric valuable, or should we rely only on logging?

- Are there other scenarios where we should consider the error-handling approach instead?

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-notes block below.
Otherwise, please describe what should be mentioned in the CHANGELOG. Use the following prefixes:
[FEATURE] [ENHANCEMENT] [PERF] [BUGFIX] [SECURITY] [CHANGE]
Refer to the existing CHANGELOG for inspiration:  https://github.com/prometheus/prometheus/blob/main/CHANGELOG.md
If you need help formulating your entries, consult the reviewer(s).
-->
```release-notes
[ENHANCEMENT] Remote write: Add logging and metrics for unexpected metadata in samples batches. When metadata entries are found in samples-only remote write batches, Prometheus now logs a warning message and increments the `prometheus_remote_storage_unexpected_metadata_total` metric to improve observability

```

cstyan reviewed (changes requested):
Sorry but counting the samples already in the batch in every `Append` call is  is not the correct way to solve the related open issue. If we're going with this kind of an implementation we should keep a running count per-batch instead.

However, looking at the related code again in `queue_manager.go` and `watcher.go`, we only ever `Append` to the `QueueManager` with `RefSample`. `RefMetadata` always goes through `StoreMetdata`. The `QueueManager` then tries to pull from `seriesMetadata` before enqueuing in data in the shards, and would enqueue a `sType` of `tSample`, `tExemplar`, or `tHistogram`.

We already have some TODO's for better handling of unexpected metadata within some `QueueManager` functions, such as in `populateV2TimeSeries`:
```
		case tMetadata:
			// TODO: log or return an error?
			// we shouldn't receive metadata type data here, it should already be inserted into the timeSeries
```

TL; DR, it seems to me like we're already not accounting for metadata within the batch size for RW2.0, so if anything we need a test that shows whether or not that is true. After that we can make a change if one is still needed.

cstyan commented on storage/remote/queue_manager_test.go:
```diff
@@ -1833,7 +1833,7 @@ func TestQueueManagerMetrics(t *testing.T) {
 	capacity := 100
 	batchSize := 10
-	queue := newQueue(batchSize, capacity)
+	queue := newQueue(batchSize, capacity, batchSize)
```
batchSize doesn't need to be passed in twice

cstyan commented on storage/remote/queue_manager.go:
```diff
@@ -1382,12 +1393,13 @@ func newQueue(batchSize, capacity int) *queue {
 
-// Append the timeSeries to the buffered batch. Returns false if it
-// cannot be added and must be retried.
+// // Append the timeSeries to the buffered batch. Returns false if it
```
double comments

cstyan commented on storage/remote/queue_manager.go:
```diff
@@ -1242,7 +1252,7 @@ func (s *shards) start(n int) {
 	newQueues := make([]*queue, n)
 	for i := 0; i < n; i++ {
-		newQueues[i] = newQueue(s.qm.cfg.MaxSamplesPerSend, s.qm.cfg.Capacity)
+		newQueues[i] = newQueue(s.qm.cfg.MaxSamplesPerSend, s.qm.cfg.Capacity, s.qm.cfg.MaxSamplesPerSend)
```
passed in twice

cstyan commented on storage/remote/queue_manager_test.go:
```diff
@@ -2362,3 +2362,152 @@ func BenchmarkBuildTimeSeries(b *testing.B) {
 }
+
+// TestRefMetadataRouting - Test with 100 samples and metadata to thoroughly check routing.
+func TestRefMetadataRouting(t *testing.T) {
```
what is this actually testing?

cstyan commented on storage/remote/queue_manager_test.go:
```diff
@@ -2362,3 +2362,152 @@ func BenchmarkBuildTimeSeries(b *testing.B) {
+	t.Log("🎯 TEST COMPLETED!")
+}
+
+func TestPopulateV2TimeSeries_UnexpectedMetadata(t *testing.T) {
```
this test isn't asserting anything either

leegin commented:
@cstyan I wanted to apologize for not cleaning up the test I initially added to check if metadata was landing in the append function. I should have removed it earlier. I had written a test to check for exponential slice capacity growth, but it turns out my test had a bug that was incorrectly showing exponential growth where none existed. Based on that flawed test, I thought we needed a fix for the capacity issue and added the maxSamples check. However, after writing a proper test case, I can now see that the new slice actually has the same capacity and there's no exponential growth happening. The existing design already handles this correctly! I've now removed those unnecessary changes.

Regarding the better handling of unexpected metadata in QueueManager functions, I noticed there was a TODO comment asking whether we should log or return an error. I've implemented the logging approach and also added a metric that increments when unexpected metadata is detected. My thinking is that this gives us better observability while being non-blocking - users get visibility into the issue without breaking their entire remote write pipeline.
I'd love to hear your thoughts on this approach! If you feel the metric isn't necessary, I'm happy to remove it. Also open to discussing whether logging vs. error handling is the right approach here, or if there's a better way to handle this altogether.

Thanks for your patience with my learning process, and sorry for the confusion with the initial test and fix!

leegin commented:
@cstyan  Thanks for the suggestion! I considered moving the metrics update inside the function, but after looking at the broader codebase patterns, I noticed that metrics updates consistently happen at the calling level rather than deep inside utility functions. To maintain consistency with this established approach, I've kept it the same way.

That said, I'm happy to make adjustments if you feel strongly about it. Also, if you think the warning log might be too noisy or not adding much value in practice, I can remove it entirely. Let me know your thoughts!

cstyan commented on storage/remote/queue_manager_test.go:
```diff
@@ -2362,3 +2362,33 @@ func BenchmarkBuildTimeSeries(b *testing.B) {
+	require.Equal(t, 0, nMetadata, "Should count 0 processed metadata")
+	require.Equal(t, 2, nUnexpected, "Should count 2 unexpected metadata")
+
+	if nUnexpected > 0 {
+		metrics.unexpectedMetadataTotal.Add(float64(nUnexpected))
+	}
+	finalCount := client_testutil.ToFloat64(metrics.unexpectedMetadataTotal)
+	require.Equal(t, initialCount+2.0, finalCount, "unexpectedMetadataTotal should be incremented by 2")
```
we don't need these lines

leegin replied:
You're right. I have removed these lines and updated the PR.

bwplotka commented:
I'm sorry, but this metric is not very useful. If anything it's a programmatic error that we should handle in a different way than surface as a production user metric. 

Can we remove this (e.g. revert this PR) and add debug log line (or just remove TODO and explicitly ignore this case?) cc @cstyan

leegin commented:
@bwplotka Just to clarify the error handling approach - are we planning to drop the entire batch when unexpected metadata is encountered, or should we handle it differently?

A few options I can think of:

- Drop entire batch - Simple fail-fast approach, forces immediate configuration fixes.
- Filter and retry - Remove metadata entries but process valid samples/exemplars/histograms

I'm leaning towards option 1 (drop entire batch) since it aligns with the "this should never happen" philosophy and provides the clearest signal that something needs to be fixed. But wanted to confirm the expected behavior before implementing, especially since dropping valid monitoring data could impact observability during misconfigurations.

bwplotka commented:
Just return an error or log line or even do nothing, without new handling - this error is extremely unlikely, and even if it happens there are no consequences (!). Plus it's for a feature that might be removed in future, that's why it's not worth investing more in this path (: 

Even no log or no error just noop is a good choice here, honestly!

Let's focus our time on more important parts and improvements (e.g. see https://github.com/prometheus/prometheus/pull/17036)

bwplotka commented:
Thanks for helping on this one - i appreciate attention to details here, but this particular code path is not worth it (:

cstyan commented:
In general I would prefer to be explicit about unexpected behaviour and notify the user in some way. Whether `populateV2TimeSeries` returns an error or logs directly I don't have a strong opinion on.

We're already not putting the seen metadata entries into the v2 write request in that function, so the write request that is built should still be valid and can be sent. 

IMO, remove metric + add debug level log line, don't consider the built write request invalid and still send it.

cstyan commented:
> If anything it's a programmatic error that we should be handled in a different way than surface as a production user metric.

This is a good point though. As the code is written right now there's no way within prometheus itself to hit this issue. Returning an error would allow us to enforce that via a test. Debug log line might still be nice, but that might only really help projects that are using our remote write code as a library.
